File: src/data/data_loader.py
"""
Data Loader
Handles loading and initial processing of labor market data
"""

import pandas as pd
import numpy as np
from pathlib import Path
from typing import Optional, Dict, Any
from ..utils.config import Config
from ..utils.logger import setup_logger
from ..utils.helpers import quarter_to_date

logger = setup_logger()

class DataLoader:
    """Handles loading and initial processing of labor market data"""
    
    def __init__(self):
        self.data = None
        self.metadata = {}
        
    def load_csv(self, file_path: str) -> pd.DataFrame:
        """
        Load data from CSV file
        
        Args:
            file_path: Path to CSV file
            
        Returns:
            Loaded DataFrame
        """
        try:
            logger.info(f"Loading data from {file_path}")
            
            # Load CSV
            df = pd.read_csv(file_path)
            
            # Validate columns
            missing_cols = set(Config.REQUIRED_COLUMNS) - set(df.columns)
            if missing_cols:
                raise ValueError(f"Missing required columns: {missing_cols}")
            
            # Convert Quarter to datetime
            df['Date'] = df['Quarter'].apply(quarter_to_date)
            
            # Sort by date
            df = df.sort_values('Date').reset_index(drop=True)
            
            # Store metadata
            self.metadata = {
                'file_path': file_path,
                'n_records': len(df),
                'start_date': df['Quarter'].iloc[0],
                'end_date': df['Quarter'].iloc[-1],
                'columns': list(df.columns)
            }
            
            self.data = df
            logger.info(f"Successfully loaded {len(df)} records")
            
            return df
            
        except Exception as e:
            logger.error(f"Error loading data: {str(e)}")
            raise
    
    def get_indicator_data(self, indicator: str) -> pd.Series:
        """
        Get data for specific indicator
        
        Args:
            indicator: Indicator name (LFPR, ER, UR, UER)
            
        Returns:
            Pandas Series with indicator data
        """
        if self.data is None:
            raise ValueError("No data loaded. Call load_csv() first.")
        
        if indicator not in Config.REQUIRED_COLUMNS[1:]:
            raise ValueError(f"Invalid indicator: {indicator}")
        
        return self.data[indicator]
    
    def get_date_range(self) -> tuple:
        """
        Get date range of loaded data
        
        Returns:
            Tuple of (start_date, end_date)
        """
        if self.data is None:
            return None, None
        
        return self.data['Quarter'].iloc[0], self.data['Quarter'].iloc[-1]
    
    def get_summary_statistics(self) -> Dict[str, Dict[str, float]]:
        """
        Calculate summary statistics for all indicators
        
        Returns:
            Dictionary of statistics for each indicator
        """
        if self.data is None:
            raise ValueError("No data loaded")
        
        stats = {}
        for indicator in Config.REQUIRED_COLUMNS[1:]:
            data = self.data[indicator]
            stats[indicator] = {
                'mean': data.mean(),
                'median': data.median(),
                'std': data.std(),
                'min': data.min(),
                'max': data.max(),
                'q25': data.quantile(0.25),
                'q75': data.quantile(0.75)
            }
        
        return stats
    
    def export_processed_data(self, output_path: Optional[str] = None) -> str:
        """
        Export processed data to CSV
        
        Args:
            output_path: Output file path (optional)
            
        Returns:
            Path to exported file
        """
        if self.data is None:
            raise ValueError("No data to export")
        
        if output_path is None:
            output_path = Config.PROCESSED_DATA_DIR / "processed_data.csv"
        
        self.data.to_csv(output_path, index=False)
        logger.info(f"Data exported to {output_path}")
        
        return str(output_path)

File: labor-market-analytics/server/services/data-loader.js
/**
 * Data Loader Service
 * Handles loading and initial processing of labor market data
 */

const fs = require("fs")
const path = require("path")
const csv = require("csv-parse/sync")

class DataLoader {
  constructor() {
    this.data = null
    this.metadata = {}
  }

  /**
   * Load data from CSV file
   * @param {string} filePath - Path to CSV file
   * @returns {Array} Loaded data array
   */
  loadCSV(filePath) {
    try {
      console.log(`[DataLoader] Loading data from ${filePath}`)

      const fileContent = fs.readFileSync(filePath, "utf-8")
      const records = csv.parse(fileContent, {
        columns: true,
        skip_empty_lines: true,
        cast: (value, context) => {
          if (context.column === "Quarter") return value
          return isNaN(value) ? value : Number.parseFloat(value)
        },
      })

      // Validate required columns
      const requiredColumns = ["Quarter", "LFPR", "ER", "UR", "UER"]
      if (records.length > 0) {
        const missingCols = requiredColumns.filter((col) => !(col in records[0]))
        if (missingCols.length > 0) {
          throw new Error(`Missing required columns: ${missingCols.join(", ")}`)
        }
      }

      // Convert Quarter to Date and sort
      const data = records
        .map((record) => ({
          ...record,
          Date: this.quarterToDate(record.Quarter),
        }))
        .sort((a, b) => new Date(a.Date) - new Date(b.Date))

      // Store metadata
      this.metadata = {
        filePath,
        nRecords: data.length,
        startDate: data[0]?.Quarter,
        endDate: data[data.length - 1]?.Quarter,
        columns: Object.keys(data[0] || {}),
      }

      this.data = data
      console.log(`[DataLoader] Successfully loaded ${data.length} records`)

      return data
    } catch (error) {
      console.error(`[DataLoader] Error loading data: ${error.message}`)
      throw error
    }
  }

  /**
   * Convert quarter string to date
   * @param {string} quarter - Quarter string (e.g., "2024 Q1")
   * @returns {string} ISO date string
   */
  quarterToDate(quarter) {
    const [year, q] = quarter.split(" Q")
    const month = (Number.parseInt(q) - 1) * 3
    return new Date(Number.parseInt(year), month, 1).toISOString().split("T")[0]
  }

  /**
   * Get data for specific indicator
   * @param {string} indicator - Indicator name
   * @returns {Array} Indicator values
   */
  getIndicatorData(indicator) {
    if (!this.data) throw new Error("No data loaded")
    return this.data.map((record) => record[indicator])
  }

  /**
   * Get summary statistics for all indicators
   * @returns {Object} Statistics for each indicator
   */
  getSummaryStatistics() {
    if (!this.data) throw new Error("No data loaded")

    const indicators = ["LFPR", "ER", "UR", "UER"]
    const stats = {}

    indicators.forEach((indicator) => {
      const values = this.getIndicatorData(indicator).filter((v) => v !== null && v !== undefined)
      values.sort((a, b) => a - b)

      stats[indicator] = {
        mean: values.reduce((a, b) => a + b, 0) / values.length,
        median: values[Math.floor(values.length / 2)],
        std: this.calculateStd(values),
        min: Math.min(...values),
        max: Math.max(...values),
        q25: values[Math.floor(values.length * 0.25)],
        q75: values[Math.floor(values.length * 0.75)],
      }
    })

    return stats
  }

  /**
   * Calculate standard deviation
   * @private
   */
  calculateStd(values) {
    const mean = values.reduce((a, b) => a + b, 0) / values.length
    const variance = values.reduce((a, b) => a + Math.pow(b - mean, 2), 0) / values.length
    return Math.sqrt(variance)
  }

  getMetadata() {
    return this.metadata
  }
}

module.exports = DataLoader
